### PR TITLE
RSE-169: Fix: Custom Replacement in Key-Value Log Filter Plugin

### PR DESCRIPTION
**Pull request by DarwisNarvaezDev**
# Fix: RSE-169 Issues with PDPAOP Data capture plugin and Ansible using force_color=true
As seen [here](https://pagerduty.atlassian.net/jira/software/c/projects/RSE/boards/716?modal=detail&selectedIssue=RSE-169&assignee=6201489c539d790069dd410f), if PDPAOP runs a command with Ansible configured to append Ansi Scape Codes to the output ***and*** we capture the data in key-values, the key value comes with the Ansi Scape Code appended.

## The problems
1. The ansi color codec for rundeck scapes the values to render the colors in the GUI but didn't remove the values from the values captured.
2. The user may not want in his keys a underscore.

## The solution
If we filter the output data with the 'invalid characters' method of the key-value plugin, we end up with an unwanted underscore in the 'key' parameter, so we extended plugin's functionality to make possible choosing the replacement for the unwanted values.

## Exhibits
Environment:
RD: Compiled WAR
Ansible -v ```2.10.8```, configured with ```force_color=true```
Command (job): ```ansible all -u <user> -m ping --private-key <key>``` 

Pre-fix:
Screenshot from 2022-10-18 17-04-18

Key-Value Data log filter definition:
Screenshot from 2022-10-18 17-04-54

Post-fix:
Screenshot from 2022-10-18 17-05-13

### CODE REVIEW

File: rundeckapp/src/main/groovy/com/dtolabs/rundeck/server/plugins/logging/SimpleDataFilterPlugin.groovy [hunk 1]
```diff
@@ -172,7 +203,17 @@ See the [Java Pattern](https://docs.oracle.com/javase/8/docs/api/java/util/regex
                 }
                 if (key && value) {
                     if(invalidKeyPattern){
-                        def validKey = key.replaceAll(invalidKeyPattern,"_")
+                        def validKey = null
+                        def emptyReplacement = invalidCharactersReplacement == null
```
Comment (ronaveva) [inline rundeckapp/src/main/groovy/com/dtolabs/rundeck/server/plugins/logging/SimpleDataFilterPlugin.groovy:207]: This should be inside if context
`                        if( replaceFilteredResult ){`

File: rundeckapp/src/test/groovy/com/dtolabs/rundeck/server/plugins/logging/SimpleDataFilterPluginSpec.groovy [hunk 2]
```diff
@@ -224,6 +224,52 @@ class SimpleDataFilterPluginSpec extends Specification {
...
+                1 * context.log(1, _)
+            }else{
+                0 * context.log(1, _)
+            }
+        }
+
+        where:
+        doWarn | validKeyPattern        | regex                    | invalidStringReplacement | name      | lines                           | expect
+        true   | '\\s|\\$|\\{|\\}|\\\\' | '^RUNDECK:DATA:(.+?)$'   | ''                       | ' ubuntu' | ['RUNDECK:DATA:zangief']        | [ubuntu: 'zangief']
+        true   | '\\s|\\$|\\{|\\}|\\\\' | '^RUNDECK:DATA:(.+?)$'   | 'Football'               | ' wimple' | ['RUNDECK:DATA:zangief']        | [Footballwimple: 'zangief']
```
Comment (ronaveva) [inline rundeckapp/src/test/groovy/com/dtolabs/rundeck/server/plugins/logging/SimpleDataFilterPluginSpec.groovy:269]: missing test for "doWarn = false"

File: rundeckapp/src/test/groovy/com/dtolabs/rundeck/server/plugins/logging/SimpleDataFilterPluginSpec.groovy [hunk 3]
```diff
@@ -224,6 +224,52 @@ class SimpleDataFilterPluginSpec extends Specification {
...
+        when:
+        plugin.init(context)
+        events.each {
+            plugin.handleEvent(context, it)
+        }
+        plugin.complete(context)
+        then:
+
+        sharedoutput.getSharedContext().getData(ContextView.global())?.getData() == (expect ? ['data': expect] : null)
+        if (expect) {
```
Comment (ronaveva) [inline rundeckapp/src/test/groovy/com/dtolabs/rundeck/server/plugins/logging/SimpleDataFilterPluginSpec.groovy:258]: unnecessary condition (?), expect is never null/false

**Review by ronaveva — CHANGES_REQUESTED** [review]
Please check the comments

File: rundeckapp/src/test/groovy/com/dtolabs/rundeck/server/plugins/logging/SimpleDataFilterPluginSpec.groovy (hunk 3 above)
Comment (DarwisNarvaezDev) [reply rundeckapp/src/test/groovy/com/dtolabs/rundeck/server/plugins/logging/SimpleDataFilterPluginSpec.groovy:258]: Done

File: rundeckapp/src/test/groovy/com/dtolabs/rundeck/server/plugins/logging/SimpleDataFilterPluginSpec.groovy (hunk 2 above)
Comment (DarwisNarvaezDev) [reply rundeckapp/src/test/groovy/com/dtolabs/rundeck/server/plugins/logging/SimpleDataFilterPluginSpec.groovy:269]: Done

File: rundeckapp/src/main/groovy/com/dtolabs/rundeck/server/plugins/logging/SimpleDataFilterPlugin.groovy (hunk 1 above)
Comment (DarwisNarvaezDev) [reply rundeckapp/src/main/groovy/com/dtolabs/rundeck/server/plugins/logging/SimpleDataFilterPlugin.groovy:207]: 'validKey' is used by different 'if' statements. (see line 215 & 218)

**Review by ronaveva — APPROVED** [review]
LGTM

**Pull request merged**